autogen/agentchat/conversable_agent.py

```diff
@@ -1328,6 +1482,41 @@ def update_function_signature(self, func_sig: Union[str, Dict], is_remove: None)
 
         self.client = OpenAIWrapper(**self.llm_config)
 
+    def update_tool_signature(self, tool_sig: Union[str, Dict], is_remove: None):
```

Comment:
Based on the [latest OpenAI API doc on function calling](https://platform.openai.com/docs/guides/function-calling), it looks like the `tools` argument is replacing the `functions` argument. Shall we combine `update_tool_signature` with `update_function_signature`? We should probably also support the [decorator-style registration]() for tools. 

@davorrunje any thoughts on this?

Reply:
yeah I think the decorator style is more recent, should definitely be supported also, will try to take a look at it.

Tool and function signatures are slightly different. General `function` functionality could be auto translated to tools definitions, removing functions from the api call. Or functions could have a depreciation method for users to explicitly switch over. But I don't think I would do a simple combination of the methods.

Could also be its own pr as a separate concern, either way

Reply:
Since the method decorator doesn't reference the "functions" naming, I'll look at just switching its functionality to use tools instead of functions. Let me know if you'd prefer otherwise

Reply:
For the decorator API implementation, replacing functions with tools is preferred in my opinion. Although I am worried that this may affect older clients. 

Yes I agree for now we can have duplicate code for this PR and have a separate PR to address deprecation issue.

Reply:
@ekzhu I agree. Replacing functions with tools should be quite straightforward, we just need to place generated JSON under "tools".

I am on vacation this week and a bit slower to respond. I'll review the proposed changes tomorrow. In any case, we can polish it in another PR.